ocaml/xapi/message_forwarding.ml

```diff
@@ -745,9 +745,27 @@ functor
         info "Pool.eject: pool = '%s'; host = '%s'"
           (current_pool_uuid ~__context)
           (host_uuid ~__context host) ;
+        let uuid = Db.Host.get_uuid __context host in
+        let master = Helpers.get_master ~__context in
+        let certs = Certificates_sync.host_certs_of ~__context host in
         let local_fn = Local.Pool.eject ~host in
+        let other =
+          Xapi_pool_helpers.get_master_slaves_list ~__context
+          |> List.filter (fun h -> h <> host && h <> master)
+        in
+        (* call eject on all other slaves first; they can't communicate
+           with host after that *)
+        other
+        |> List.iter (fun h ->
+               do_op_on ~local_fn ~__context ~host:h (fun session_id rpc ->
+                   Client.Pool.eject rpc session_id host)) ;
```

Comment:
This should be done _after_ the host has self-ejected. The ejection may fail when HA is enabled or there's a VM resident in this host. Is there anything stopping this from being done after the ejection?

Reply:
> they can't communicate with host after that

They don't need to communicate with the ejected host. Only the master does, and only to get the outcome of the rejection.

Reply:
If the ejection fails the hosts will not have the certificate to contact the host. If, for some reason the master rotates, it won't be able to forward calls to the host that was tried to be ejected.

Reply:
If the host is ejected first, it is impossible to later refer to it by its ref afterwards, as is required when calling `Xapi_pool.eject` on all other hosts.

Reply:
If the cleanup of the remaining host needs to happen after the host was ejected it needs a new internal API call that refers to the host by its UUID. This is ugly but avoids the worse problem what happens if eject fails.

Reply:
I see, it's not possible to change the ref to a uuid because it's part of the API.
In that case would it be possible to  separate the pre_checks that can make the eject to fail into another function so they can be run before deleting certificates?

Otherwise somehow exposing the function `eject_certs_from_fs_for` to be able to retrieve the uuid beforehand and do the file-system after the ejectee has rebooted.

Reply:
We could delay the removal from the database and only do it once:
* the host ejected itself
* all other hosts removed the certs
This would allows us to use the API call using the ref